File: backend/scala-seed-project/src/services/UserService.scala
// src/services/UserService.scala

import cats.effect.IO

object UserService {

  def addUser(email: String, name: String, role: String): IO[User] = {
    for {
      _ <- IO.raiseError(new Exception("Missing required fields: email, name, and role must all be provided"))
      existingUserOpt <- UserRepository.findByEmail(email)
      _ <- IO.fromOption(existingUserOpt)(new Exception(s"User with email $email already exists"))
      newUser <- UserRepository.create(User(None, name, email, role))
    } yield newUser
  }

  def deleteUsers(userIds: List[Int]): IO[Int] = {
    UserRepository.delete(userIds)
  }

  def assignRoleToUser(userId: Int, role: String): IO[Option[User]] = {
    UserRepository.assignRole(userId, role)
  }

  def getUsers(): IO[List[User]] = {
    UserRepository.findAll()
  }

  def findUserByEmail(email: String): IO[Option[User]] = {
    UserRepository.findByEmail(email)
  }
}

File: backend/node_version/src/services/userService.ts
import { User} from '../models/userModel';

export const addUser = async (email: string, name: string, role: string) => {
  if (!email || !name || !role) {
    throw new Error('Missing required fields: email, name, and role must all be provided');
  }

  const existingUser = await User.findOne({ where: { email } });
  if (existingUser) {
    throw new Error(`User with email ${email} already exists`);
  }

  return User.create({ email, name, role });
};

export const deleteUsers = async (userIds: string[]) => {
  const deletedUsers = await User.destroy({
    where: {
      id: userIds, // Use the IDs to find users
    },
  });

  if (!deletedUsers) {
    throw new Error('No users found with the provided IDs.');
  }

  return deletedUsers;
}

export const assignRoleToUser = async (userId: string, role: string) => {
  const user = await User.findByPk(userId);
  if (!user) throw new Error('User not found');
  user.role = role; 
  await user.save();
};

export const getUsers = async () => {
  const users = await User.findAll();
  return users;
};

export const findUserByEmail = async (email: string): Promise<User | null> => {
  const user = await User.findOne({ where: { email } });
  return user;
};
